mixnet/mixnet.py

```diff
@@ -12,31 +16,105 @@
 class Mixnet:
     mix_nodes: List[MixNode]
 
-    # Build a new topology deterministically using an entropy.
-    # The entropy is expected to be injected from outside.
-    #
-    # TODO: Implement constructing a new topology in advance to minimize the topology transition time.
-    #       https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#9a7f6089e210454bb11fe1c10fceff68
-    def build_topology(
+    def __init__(
         self,
-        entropy: bytes,
+        mix_nodes: List[MixNode],
         n_layers: int,
         n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+        initial_entropy: bytes,
+    ):
+        self.mix_nodes = mix_nodes
+
+        self.set_entropy(initial_entropy)
+
+        self.topology_updater = MixnetTopologyUpdater(
+            mix_nodes, n_layers, n_nodes_per_layer, mixnet_epoch_sec, proactive_sec
+        )
+        self.topology_updater.daemon = True
+        self.topology_updater.start()
+
+    def set_entropy(self, entropy: bytes) -> None:
+        """
+        Set entropy (seed) to the Fisher-Yates shuffler
+        that is going to be used for deterministic topology construction.
+
+        This method is expected to be called periodically by the user.
+        Until the next entropy is provides, Fisher-Yates shuffler continues to works
+        based on this entropy set once.
+        """
+        FisherYates.set_seed(entropy)
+
+    def choose_mixnode(self) -> MixNode:
+        return random.choice(self.mix_nodes)
+
+    def current_topology(self) -> MixnetTopology:
+        return self.topology_updater.current_topology()
+
+
+class MixnetTopologyUpdater(Thread):
+    def __init__(
+        self,
+        mix_nodes: List[MixNode],
+        n_layers: int,
+        n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+    ):
+        super().__init__()
+        self.mix_nodes = mix_nodes
+        self.n_layers = n_layers
+        self.n_nodes_per_layer = n_nodes_per_layer
+        self.mixnet_epoch_sec = mixnet_epoch_sec
+        self.proactive_sec = proactive_sec
+
+        self.lock = threading.Lock()
+        self.topology = self.build_topology()
+        self.next_topology = None
+
+    def run(self) -> None:
+        next_epoch_ts = datetime.now() + timedelta(seconds=self.mixnet_epoch_sec)
+
+        while True:
+            time.sleep(1 / 1000)
+
+            now = datetime.now()
+            if now < next_epoch_ts - timedelta(seconds=self.proactive_sec):
+                continue
+
+            if self.next_topology is None:
+                self.next_topology = self.build_topology()
```

Comment:
If I understand it correctly, in this version, we are refreshing the topology only on once per epoch?

Reply:
Yes. The topology is updated once per `mixnet_epoch` (e.g. 30min), which is going to be shorter than `consensus_epoch` (e.g. 5d). Also, the `mixnet_epoch` is going to be independent with `consensus_slot` (20sec). That might be longer.

Reply:
I think that it might be easier to think about mixnet functions as a tool that will be used by the consensus. In my head, the consensus should be responsible for triggering mixnet function, as the consensus should be treated as the source of truth. That is, what if we focus only on providing an API that potentially could be used as part of the consensus who will be responsible for synchronization of time and data, and the mixnet API will only respond (by setting the connectivity) accordingly to data provided by consensus. This way we could simplify a lot the internal logic of the mixnet and relieve it from knowing anything about the time or consensus.

The API could be very simple, such like:
`topology = get_mixnet_topology(`nodes_list`, entropy)` <-- returns topology configuration based on the entropy and the list of nodes.
`set_mixnet_topology(topology)` <-- activates the topology.

I think that mixnet should only care about the two above, and all synchronization/configuration issues should be abstracted from the mixnet and done on the consensus side. Therefore, the consensus be responsible for calculating and providing synchronized data in a timely manner to the mixnet.

Technically, I think that there should be another layer on top of the consensus that would do the above in conjunction to parsing governance transactions (coordination layer?), through which some of the global parameters will be set. But we don't have it yet so I'm using consensus here.

Reply:
> That is, what if we focus only on providing an API that potentially could be used as part of the consensus who will be responsible for synchronization of time and data,

Exactly. This is the same as what I thought at https://github.com/logos-co/nomos-specs/pull/51#discussion_r1458613901. But, the two APIs that you suggested are the more abstracted version than my recent fix https://github.com/logos-co/nomos-specs/commit/9a64242f4596f9b5d78aa103ad002d8394b1ad5f (linked in in https://github.com/logos-co/nomos-specs/pull/51#discussion_r1458613901). In general, I prefer your API because the mixnet doesn't need to care about the `entropy_delay_sec` (in my recent fix).

The last thing we need to take into account when designing these two APIs is that we also want for the mix node to establish connections with adjacent mix layer in the new topology [in advance](https://github.com/logos-co/nomos-specs/commit/9a64242f4596f9b5d78aa103ad002d8394b1ad5f#diff-4b6fc539badd8395558d0df89c79dfa602639423cb4b9f91593249c733a4ba6eR82). Maybe we can design mix node to start connection establishment in background as soon as the new topology is returned from `get_mixnet_topology` API. Unless the consensus layer calls `set_mixnet_topology` in very short time, the mix node can have time to establish connections in advance.

Reply:
I'm not so sure this should be like that. Consensus should know nothing of the comunication layer behind. It just knows there is something where it can send messages and receive messages. Eventually the mixnet can be even opt-in for that and should not matter. This being said, I think consensus can trigger some events of what is happening in the system (we get to decide that) then mixnet can use of that. But as possible, consensus and mixnet is better they do not know each other or the do know as little as possible.

Reply:
From https://github.com/logos-co/nomos-specs/pull/51#discussion_r1458613901, I removed the concept `mixnet_epoch`. But previously, yes, mixnet epoch was much shorter than consensus epoch. 

Now, I agree that having mixnet epoch is little complicated. Instead we can remove mixnet epoch, and make mixnet update topology whenever a new entropy is injected from the upper layer. Even, mixnet doesn't need to care about the frequency of entropy injection. Also, consensus layer doesn't need to know deeply about what would happen in the mixnet. The one thing that consensus layer has to do is propagating the entropy updated by it to other layers, which is something like a logical clock, as you said.

I think this decouples mixnet and consensus quite well, and this is what https://github.com/logos-co/nomos-specs/commit/9a64242f4596f9b5d78aa103ad002d8394b1ad5f and https://github.com/logos-co/nomos-specs/pull/51#discussion_r1459185382 try to achieve.
I think that you, Marcin, and I am already on the same page with this point. Did I understand your comment correctly?

Reply:
@madxor @danielSanchezQ I think we're almost on the same page and we can now come to a conclusion.

I made a commit 9a64242f4596f9b5d78aa103ad002d8394b1ad5f for clarity and another commit 1a970641a8847f8695e4c3f0cafb48ab32c3971f to rename a variable. These commit has API different from what Marcin suggested in https://github.com/logos-co/nomos-specs/pull/51#discussion_r1459185382, but I'm going to explain about this at the end of this comment.

In these commit, the consensus layer just injects a new entropy when it thinks that the new entropy is almost safe to use. Potentially, the consensus layer will be able to have some delays before injecting the entropy, instead of injecting it as soon as the random beacon is updated, because of https://github.com/logos-co/nomos-specs/pull/51#discussion_r1457694044. But anyway, it's the responsibility of the consensus layer. Mixnet layer doesn't need to know about this detail.

In perspective of mixnet, it just uses a new entropy whenever it is injected. Mixnet doesn't need to care about the frequency of entropy injection. It is up to the consensus layer. Even if the new entropy is injected every 20 seconds, it's okay. Mixnet will just use it to build a new topology, as soon as it arrives in. 

The one thing Mixnet cares about is putting some delay before switching to the new topology. As mentioned in https://github.com/logos-co/nomos-specs/pull/51#discussion_r1459244787, it takes some time for each mix node to establish new connections according to the new topology. Also, there may be some timing differences on the new entropy being injected. Of course, we can ignore this, by allowing a chaotic period in which packets are propagated through unexpected paths. But, I think we can minimize this chaotic period by putting some delay (e.g. 1min) before switching to the new topology. Then, each mix node can earn 1min to calculate the new topology and establish new connections. This is what 1a970641a8847f8695e4c3f0cafb48ab32c3971f tries to describe. This delay is not perfect, but at least, it can minimize the chaotic transition in the network that cannot be fully synchronized.

As I mentioned above, I designed API different from what Marcin suggested, because I didn't want to expose the concept of "topology" to the consensus layer by providing `get/set_topology` API. In that case, the consensus layer may be confused about how long it should hold the `topology` returned from `get_topology(entropy)` before calling `set_topology(topology)`. Instead of that, I thought it's simpler to have a single API `inject_entropy(entropy)`, so that the consensus layer just injects a new entropy whenever it wants and forgets about that.

What do you think about this? We have had many valuable discussions, and I think we are in the end of the discussion.

Reply:
I'm just attaching a diagram that helps our discussion. The option 2 is what my code is suggesting, but let's discuss today.
![mixnet-topology-update drawio](https://github.com/logos-co/nomos-specs/assets/5462944/fff7b798-02bd-43bb-82b1-5f963614ded7)
Even if we choose the option 2, the first topology generation (when the process is started) should be like the option 1 probably.

Reply:
I think that we are on the right path. However, I think we are still missing another layer to make it perfect. I fully agree that there should be a logical separation between the mixnet (and its time-awareness) and consensus (and its network awareness). Therefore, it might be necessary to create another layer on top of the mixnet and consensus, we can call it robustness (or privacy) agent/layer (or something). The robustness layer would be responsible for reading consensus and setting mixnet, we could configure it separately from the mixnet and consensus. Therefore, the flow would always be from the consensus event to mixnet action with independently configured logic.

For example, if change the length of the consensus epoch or slot time, then we should not adjust mixnet delay, that should be taken care by the robustness abstraction, which would translate the consensus timings to the mixnet ones (in practice, I imagine that mixnet will not be aware on any specific timings but will react on robustness calls).

Reply:
I agree as well. We need something in the middle to coordinate efforts. I wan to understand more about the delays as I am getting a bit lost on that. I feel the responsibility is kind of mixed. This system should be totally reactive IMO.

Reply:
This is a summary of our discussion today. 

![mixnet-topology-update drawio(3)](https://github.com/logos-co/nomos-specs/assets/5462944/4b923d9a-9d55-40f0-aed3-f2fd442c19e1)

- We're gonna have a separate "robustness" layer on top of consensus layer and mixnet layer.
- Robustness layer is gonna know parameters for building mixnet topologies.
- Whenever a new entropy is stabilized in the consensus layer, the consensus layer is going to emit the new entropy to the robustness layer, along with the updated list of nodes in the network. In general, this is going to happen once per consensus epoch, but it depends on the design of consensus layer, strictly speaking.
- As soon as a new information arrives in, the robustness layer "immediately" builds a new topology configuration based on the current parameter that it knows. The topology configuration means the list of mix nodes for each layer (for now in v1).
- As soon as the new topology configuration is generated, the robustness layer injects it to the mixnet layer. There is going to be no time delay between those two events.
- The mixnet layer switches the topology configuration immediately, and starts establishing new QUIC connections in background.
  - Packets based on the new topology can arrive in the mix node, even before all new connections are established. But, it is totally okay. Mix node always establish a new connection on demand to forward the packet with the best effort. Even if packets based on unexpected topology is arrived in, mix node never reject packets. This design is reasonable because we assume that all nodes in the network are ready to act as a mix node (at least for v1).